nv_peer_mem.c

```diff
@@ -187,18 +239,58 @@ static int nv_dma_map(struct sg_table *sg_head, void *context,
 		(struct nv_mem_context *) context;
 	struct nvidia_p2p_page_table *page_table = nv_mem_context->page_table;
 
-	if (nv_mem_context->page_table->page_size != NVIDIA_P2P_PAGE_SIZE_64KB) {
+	if (page_table->page_size != NVIDIA_P2P_PAGE_SIZE_64KB) {
 		peer_err("nv_dma_map -- assumption of 64KB pages failed size_id=%u\n",
 					nv_mem_context->page_table->page_size);
 		return -EINVAL;
 	}
 
+#if NV_DMA_MAPPING
+	{
+		struct nvidia_p2p_dma_mapping *dma_mapping;
+		struct pci_dev *pdev = to_pci_dev(dma_device);
+
+		if (!pdev) {
+			peer_err("nv_dma_map -- invalid pci_dev\n");
+			return -EINVAL;
+		}
+
+		ret = nvidia_p2p_dma_map_pages(pdev, page_table, &dma_mapping);
+		if (ret) {
+			peer_err("nv_dma_map -- error %d while calling nvidia_p2p_dma_map_pages()\n", ret);
+			return ret;
+		}
+
+		if (!NVIDIA_P2P_DMA_MAPPING_VERSION_COMPATIBLE(dma_mapping)) {
+			peer_err("error, incompatible dma mapping version 0x%08x\n",
+				 dma_mapping->version);
+			nvidia_p2p_dma_unmap_pages(pdev, page_table, dma_mapping);
+			return -EINVAL;
+		}
+
+		nv_mem_context->npages = dma_mapping->entries;
+
+		ret = sg_alloc_table(sg_head, dma_mapping->entries, GFP_KERNEL);
```

Comment:
On success, nv_mem_put_pages frees the memory otherwise the sg_alloc_table frees any leftovers